### DISCUSSION TITLE
quality time on kubernetes/openshift in corporate environments

### DISCUSSION THREAD

**Discussion by zubcevic** [ICTU/quality-time#4741] [General]
I have looked at _quality time_ and tried to set it up on a corporate on-premise openshift environment.

(Question) I wonder where/how I could share the instructions to set this up. And perhaps discuss stuff that I encountered in that process.

Since the current instructions rely on docker-compose, you need some adjustments to run it on openshift.

Summary of the steps involved:

1. Use kompose to transform the docker compose file to a Helm chart
2. Rename some elements to conform to openshift/corporate naming policies in the Helm chart
3. Pull images using some corporate container proxy
4. Rebuild images to add corporate certificate authorities to allow to connect to corporate cicd tools
5. Rename the base images in the Helm charts to point to these custom images
6. Add K8S/openshift service yaml instructions for the port definitions
7. Add OpenShift route for the overall entrypoint
8. Adjust the K8S persistent volume claim to what is available
9. Adjust environment variables (like LDAP) to point to the corporate LDAP 

Issues encountered:

1. NGNIX www/proxy image will not work on a corporate strict OpenShift environment. The Red Hat UBI8 NGNIX will work. So this build has been done based on the new base image and a similar NGNIX config
2. Some directories are not accessible for random non-root users. In the rebuild of the images this can be corrected
3. HTTPS issues. OS level certificates have been added + a Python specific environment variable that points to a certificate bundle
4. (open) SonarQube integration works but it seems to select rulesets that are very specific. Can these be configured somewhere?
5. (open) SonarQube integration complains about the version conventions used. 1.0.0-SNAPSHOT e.g. will fail. Can these be configured somewhere?
6. GitLab is supported, Bitbucket unfortunately not. Is there a manual to be able to add this/ build this using an upstream pull request?

So the good news is, that with some work, it will run fine in OpenShift. But it would benefit others if some of these steps could be shared somewhere

**Comment by fniessink** [comment]
Hi René, thanks for your experience report. We would welcome PRs! I think one PR could indeed be to add the steps into a chapter in the deployment document. For example, change the `docs/src/deployment.md` document to have two chapters, one with the current information about docker-compose and a new second chapter for deploying on Openshift.

With respect to your issues:

1. I guess this could be incorporated in the PR for the deployment documentation?
2. I'm not sure which directories in which containers you are referring to?
3.  This may be suitable for a separate PR?
4. The rulesets used for some specific metric (for example long methods, too many parameters) are [documented in the data model](https://github.com/ICTU/quality-time/blob/master/components/shared_data_model/src/shared_data_model/sources/sonarqube.py) but not currently configurable by end-users. We used to have functionality for this, but actually removed it because people did not use it in practice.
5. For the [source version metric](https://quality-time.readthedocs.io/en/latest/reference.html#source-version) Quality-time uses a specific library to parse version numbers. It could well be that 1.0.0-SNAPSHOT is not considered to be a valid version number by that library. Please consider [opening a new issue](https://github.com/ICTU/quality-time/issues/new/choose) for this problem.
6. The [developer documentation](https://quality-time.readthedocs.io/en/latest/development.html) has information about [adding new metrics and sources](https://quality-time.readthedocs.io/en/latest/development.html#adding-metrics-and-sources).

**Comment by zubcevic** [comment]
see #4764

**Comment by makeitstatic** [comment]
I need to do these steps similiarly for my Podman rootless deployment on RHEL8. But these are so vague I am not sure how to do all of it.
Is anyone available to consult about it?
